### DISCUSSION TITLE
Self managed Connection timeout

### DISCUSSION THREAD

**Discussion by RickIncentro** [camunda-community-hub/pyzeebe#461] [Q&A] [answered]
@dimastbk 
Thanks for continuing the work on the library. I am currently using it with a self-managed instance. In order to use it, I created an extra function (updated from create_camunda_cloud_channel). It times out after 5 min however: 
  raise _create_rpc_error(
 grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
 status = StatusCode.UNAUTHENTICATED
    details = “Failed to parse bearer token, see cause for details”
Any idea on how to efficiently refresh the token within the library? Thanks in advance!

`
from typing import Dict, Optional

import grpc
from oauthlib import oauth2
from requests import HTTPError
from requests_oauthlib import OAuth2Session

from pyzeebe.channel.channel_options import get_channel_options
from pyzeebe.errors import (
    InvalidCamundaCloudCredentialsError,
    InvalidOAuthCredentialsError,
)

def create_self_managed_channel(
    client_id: str,
    client_secret: str,
    cluster_url: str,
    channel_options: Optional[Dict] = None,
) -> grpc.aio.Channel:
    """
    Create channel connected to a Camunda Cloud cluster

    Args:
        client_id (str): The client id provided by Camunda Cloud
        client_secret (str): The client secret provided by Camunda Cloud
        cluster_id (str): The zeebe cluster id to connect to
        region (str): The cluster's region. Defaults to bru-2
        channel_options (Optional[Dict], optional): GRPC channel options. See https://grpc.github.io/grpc/python/glossary.html

    Returns:
        grpc.aio.Channel: A GRPC Channel connected to the Zeebe gateway.

    Raises:
        InvalidCamundaCloudCredentialsError: One of the provided camunda credentials is not correct
    """
    channel_credentials = _create_camunda_self_managed_credentials(client_id, client_secret, cluster_url)

    return grpc.aio.secure_channel(
        cluster_url,
        channel_credentials,
        options=get_channel_options(channel_options),
    )

def _create_camunda_self_managed_credentials(
    client_id: str, client_secret: str, cluster_url: str
) -> grpc.ChannelCredentials:
    try:
        access_token = _get_access_token(
            "URL",
            client_id,
            client_secret,
            "URL",
        )
        return _create_oauth_credentials(access_token)
    except InvalidOAuthCredentialsError as oauth_error:
        raise InvalidCamundaCloudCredentialsError(client_id, cluster_url) from oauth_error

def _get_access_token(url: str, client_id: str, client_secret: str, audience: str) -> str:
    try:
        client = oauth2.BackendApplicationClient(client_id)
        client.prepare_request_body(include_client_id=True)
        with OAuth2Session(client=client) as session:
            response = session.post(
                url,
                data={
                    "client_id": client_id,
                    "client_secret": client_secret,
                    "grant_type": "client_credentials"
                },
            )
            response.raise_for_status()
            return response.json()["access_token"]
    except HTTPError as http_error:
        raise InvalidOAuthCredentialsError(url=url, client_id=client_id, audience=audience) from http_error

def _create_oauth_credentials(access_token: str) -> grpc.ChannelCredentials:
    token_credentials = grpc.access_token_call_credentials(access_token)
    ssl_credentials = grpc.ssl_channel_credentials()
    return grpc.composite_channel_credentials(ssl_credentials, token_credentials)

`

**Comment by dimastbk** [comment, marked as answer]
Hi! For self-managed Camunda Cluster you can use CamundaIdentityCredentials (see example in https://github.com/camunda-community-hub/pyzeebe/blob/master/docs/channels.rst under "Example with oauth2 (like Camunda Identity):") . This GRPC plugin can automatically update the token every five minutes (default in Camunda Self-Managed). (Available only in 4.*, that under development.)

**Comment by RickIncentro** [comment]
Thanks for your reply! However I get this error after it tries to connect multiple times:  File "c:\Users\rickb\Desktop\pyzeebe4.0\pyzeebes4.0\pyzeebe\pyzeebe\grpc_internals\zeebe_job_adapter.py", line 41, in activate_jobs        
    async for response in self._gateway_stub.ActivateJobs(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\grpc\aio\_channel.py", line 168, in __call__
    call = UnaryStreamCall(request, deadline, metadata, credentials,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\grpc\aio\_call.py", line 552, in __init__
    channel.call(method, deadline, credentials, wait_for_ready),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "src\python\grpcio\grpc\_cython\_cygrpc/aio/channel.pyx.pxi", line 124, in grpc._cython.cygrpc.AioChannel.call
grpc._cython.cygrpc.UsageError: Channel is closed.

any idea why?

**Reply by RickIncentro** [reply]
nvm it is working, thanks!
